xmppserver/src/main/java/org/jivesoftware/openfire/event/ServerSessionEventDispatcher.java

```diff
@@ -0,0 +1,133 @@
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.jivesoftware.openfire.session.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Dispatches session events (s2s). Each event has a {@link EventType type}
+ *
+ * @author Manasse Ngudia
+ */
+public class ServerSessionEventDispatcher {
+
+    private static final Logger Log = LoggerFactory.getLogger(ServerSessionEventDispatcher.class);
+
+    private static List<ServerSessionEventListener> listeners =
+            new CopyOnWriteArrayList<>();
+
+    private ServerSessionEventDispatcher() {
+        // Not instantiable.
+    }
+
+    /**
+     * Registers a listener to receive events.
+     *
+     * @param listener the listener.
+     */
+    public static void addListener(ServerSessionEventListener listener) {
+        if (listener == null) {
+            throw new NullPointerException();
+        }
+        listeners.add(listener);
+    }
+
+    /**
+     * Unregisters a listener to receive events.
+     *
+     * @param listener the listener.
+     */
+    public static void removeListener(ServerSessionEventListener listener) {
+        listeners.remove(listener);
+    }
+
+    /**
+     * Dispatches an event to all listeners.
+     *
+     * @param session the session.
+     * @param eventType the event type.
+     */
+    public static void dispatchEvent(Session session, EventType eventType) {
+        for (ServerSessionEventListener listener : listeners) {
+            try {
+                switch (eventType) {
+                    case session_created: {
+                        listener.sessionCreated(session);
+                        break;
+                    }
+                    case session_destroyed: {
+                        listener.sessionDestroyed(session);
+                        break;
+                    }
+                    case anonymous_session_created: {
+                      listener.anonymousSessionCreated(session);
+                      break;
+                    }
+                    case anonymous_session_destroyed: {
+                      listener.anonymousSessionDestroyed(session);
+                      break;
+                    }
+                    case resource_bound: {
+                      listener.resourceBound(session);
+                      break;
+                    }
+                    default:
+                        break;
+                }
+            }
+            catch (Exception e) {
+                Log.error(e.getMessage(), e);
+            }
+        }
+    }
+
+    /**
+     * Represents valid event types.
+     */
+    public enum EventType {
+
+        /**
+         * A session was created.
+         */
+        session_created,
+
+        /**
+         * A session was destroyed
+         */
+        session_destroyed,
+        
+        /**
+         * An anonymous session was created.
+         */
+        anonymous_session_created,
```

Comment:
For S2S, I don't think we have anonymous sessions, nor resource binding. We might have some other events, but for now, lets just use "session_created" and "session_destroyed".